flowmachine/flowmachine/features/subscriber/handset_stats.py

```diff
@@ -0,0 +1,196 @@
+    order to take the different usage periods into account we must calculate
+    weighted statistics of the desired characteristics. As such, here we
+    calculate the weighted characteristic, weighted by the number of seconds a
+    subscriber held that handset.  Given that we only learn about changes in
+    handset when an event occurs, this average will be biased.  For instance,
+    if a subscriber changes his handset straight after an event occur, we will
+    be tracking the handset that the subscriber used to handle the event and
+    not the handset he actually switched to. We will only learn about the new
+    handset once a new event occurs.
+
+    We further assume that the handset held before the first observed event in
+    the series is equal to the handset of the first event. Likewise, we assume
+    that the handset held after the last observed event in the series is the
+    handset of the last event.
+
+    Parameters
+    ----------
+    start, stop : str
+         iso-format start and stop datetimes
+    characteristic: {"width", "height", "depth", "weight", "display_width", "display_height"}
+        Numeric handset characteristics present in the TAC table.
+    statistic : {'count', 'sum', 'avg', 'max', 'min', 'median', 'mode', 'stddev', 'variance'}, default 'sum'
+        Defaults to sum, aggregation statistic over the durations.
+    hours : 2-tuple of floats, default 'all'
+        Restrict the analysis to only a certain set
+        of hours within each day.
+    subscriber_identifier : {'msisdn', 'imei'}, default 'msisdn'
+        Either msisdn, or imei, the column that identifies the subscriber.
+    subscriber_subset : str, list, flowmachine.core.Query, flowmachine.core.Table, default None
+        If provided, string or list of string which are msisdn or imeis to limit
+        results to; or, a query or table which has a column with a name matching
+        subscriber_identifier (typically, msisdn), to limit results to.
+
+    Examples
+    --------
+
+    >>> s = HandsetStats("2016-01-01", "2016-01-08", "width")
+    >>> s.get_dataframe()
+    """
+
+    def __init__(
+        self,
+        start,
+        stop,
+        characteristic,
+        statistic="avg",
+        *,
+        subscriber_identifier="msisdn",
+        hours="all",
+        subscriber_subset=None,
+        tables="all",
+    ):
+        self.start = start
+        self.stop = stop
+        self.characteristic = characteristic.lower()
+        self.statistic = statistic.lower()
+
+        if self.statistic not in valid_stats:
+            raise ValueError(
+                f"{self.statistic} is not a valid statistic. Use one of {valid_stats}"
+            )
+
+        if self.characteristic not in valid_characteristics:
+            raise ValueError(
+                f"{self.characteristic} is not a valid characteristic. Use one of {valid_characteristics}"
+            )
+
+        self.handsets_query = SubscriberHandsets(
```

Comment:
I know I sound like a broken record, but why not take this as an argument?

Reply:
Ok, made this as argument.